ui/app/services/api.ts

```diff
@@ -0,0 +1,185 @@
+      init.headers = headers;
+      init.body = JSON.stringify({ token });
+    }
+    return { url: newUrl, init };
+  };
+
+  setHeaders = async (context: RequestContext) => {
+    const { url, init } = context;
+    const headers = new Headers(init.headers);
+    // unauthenticated or clientToken requests should set the header in initOverrides
+    // unauthenticated value should be empty string, not undefined or null
+    if (!headers.has('X-Vault-Token')) {
+      headers.set('X-Vault-Token', this.authService.currentToken);
+    }
+    if (init.method === 'PATCH') {
+      headers.set('Content-Type', 'application/merge-patch+json');
+    }
+    // use initOverrides to set the namespace header to something other than path set in the namespace service
+    // for requests that must be made to root namespace pass empty string as value
+    const namespace = this.namespaceService.path;
+    if (!headers.has('X-Vault-Namespace') && namespace) {
+      headers.set('X-Vault-Namespace', namespace);
+    }
+
+    init.headers = headers;
+    return { url, init };
+  };
+
+  // -- Post Request Middleware --
+  showWarnings = async (context: ResponseContext) => {
+    const response = context.response.clone();
+    const json = await response?.json();
+    // currently there is only 1 endpoint that intentionally hides warnings
+    // handle that here for now, but if more endpoints are added, look for a more scalable pattern
+    const hideWarnings = context.url === '/v1/sys/internal/counters/activity';
+
+    if (json?.warnings && !hideWarnings) {
+      json.warnings.forEach((message: string) => {
+        this.flashMessages.info(message);
+      });
+    }
+  };
+
+  deleteControlGroupToken = async (context: ResponseContext) => {
+    const { url } = context;
+    const controlGroupToken = this.controlGroup.tokenForUrl(url);
+    if (controlGroupToken) {
+      this.controlGroup.deleteControlGroupToken(controlGroupToken.accessor);
+    }
+  };
+
+  formatErrorResponse = async (context: ResponseContext) => {
+    const response = context.response.clone();
+    const { headers, status, statusText } = response;
+
+    // backwards compatibility with Ember Data
+    if (status >= 400) {
+      const error: ApiError = (await response?.json()) || {};
+      error.httpStatus = response?.status;
+      error.path = context.url;
+      // typically the Vault API error response looks like { errors: ['some error message'] }
+      // but sometimes (eg RespondWithStatusCode) it's { data: { error: 'some error message' } }
+      if (error?.data?.error && !error.errors) {
+        // normalize the errors from RespondWithStatusCode
+        error.errors = [error.data.error];
+      }
+      return new Response(JSON.stringify(error), { headers, status, statusText });
+    }
+
+    return;
+  };
+
+  // the responses in the OpenAPI spec don't account for the return values to be under the 'data' key
+  // return the data rather than the entire response
```

Comment:
So does this mean we're just returning what would be inside the `data` block here?

Reply:
Yes that's correct. I'm working on a follow up commit that will address the need for the full response for void requests but I needed to update the generator for that.

The reason why this is necessary is because the OpenAPI spec has no awareness of the request parameters outside of the data key so it's expecting those to be at the top level of the response object.